src/mixins/yield-erc/ERC20Vault.sol

```diff
@@ -0,0 +1,149 @@
+import {FixedPointMathLib} from "../../utils/FixedPointMathLib.sol";
+import {SafeTransferLib} from "../../utils/SafeTransferLib.sol";
+
+/// @title Yield Bearing Vault
+/// @author joeysantoro, Transmissions11 and JetJadeja
+contract ERC20Vault is ERC20 {
+    using SafeTransferLib for ERC20;
+    using FixedPointMathLib for uint256;
+
+    /*///////////////////////////////////////////////////////////////
+                                IMMUTABLES
+    //////////////////////////////////////////////////////////////*/
+
+    /// @notice The underlying token the Vault accepts.
+    ERC20 public immutable underlying;
+
+    /// @notice The base unit of the underlying token and hence vault.
+    /// @dev Equal to 10 ** decimals. Used for fixed point arithmetic.
+    uint256 public immutable baseUnit;
+
+    /// @notice Creates a new Vault that accepts a specific underlying token.
+    /// @param _underlying The ERC20 compliant token the Vault should accept.
+    /// @param _name The name for the vault token.
+    /// @param _symbol The symbol for the vault token.
+
+    constructor(
+        ERC20 _underlying,
+        string memory _name,
+        string memory _symbol
+    )
+        ERC20(
+            _name,
+            _symbol,
+            _underlying.decimals()
+        )
+    {
+        underlying = _underlying;
+
+        baseUnit = 10**decimals;
+    }
+
+    /*///////////////////////////////////////////////////////////////
+                        DEPOSIT/WITHDRAWAL LOGIC
+    //////////////////////////////////////////////////////////////*/
+
+    /// @notice Emitted after a successful deposit.
+    /// @param user The address that deposited into the Vault.
+    /// @param underlyingAmount The amount of underlying tokens that were deposited.
+    event Deposit(address indexed user, uint256 underlyingAmount);
+
+    /// @notice Emitted after a successful withdrawal.
+    /// @param owner The address that withdrew from the Vault.
+    /// @param to The destination for withdrawn tokens.
+    /// @param underlyingAmount The amount of underlying tokens that were withdrawn.
+    event Withdraw(address indexed owner, address indexed to, uint256 underlyingAmount);
+
+    /// @notice Deposit a specific amount of underlying tokens.
+    /// @param to The address to receive shares corresponding to the deposit
+    /// @param underlyingAmount The amount of the underlying token to deposit.
+    function deposit(address to, uint256 underlyingAmount) external virtual returns (uint256 shares) {
+        shares = underlyingAmount.fdiv(exchangeRate(), baseUnit);
+        // Determine the equivalent amount of shares and mint them.
+        _mint(to, shares);
+
+        emit Deposit(to, underlyingAmount);
+
+        // Transfer in underlying tokens from the user.
+        // This will revert if the user does not have the amount specified.
+        underlying.safeTransferFrom(msg.sender, address(this), underlyingAmount);
```

Comment:
Consider tracking the balance of underlying in the contract, and here subtract any surplus from the transfer, potentially even not executing the transfer if not needed.

This will allow other contracts that produce or move underlying to leave it at this vault, executing `deposit` afterwards. This will remove the need for an `approve` (less gas) and make integrations easier.

For examples of using this pattern, check the yield contracts.

Reply:
I love this suggestion. For Fei PCV deposits we use an "assume tokens were already transferred model"

Reply:
> This will allow other contracts that produce or move underlying to leave it at this vault, executing deposit **_afterwards_**.

probably self-evident, but this would have to happen atomically right? Otherwise, I can see some sniping happening

Reply:
yeah would be atomic. After discussing with @transmissions11 going to leave it approval only for this base implementation.